analyzer/src/traversal/functions.rs

```diff
@@ -351,31 +351,37 @@ fn emit(
     stmt: &Node<fe::FuncStmt>,
 ) -> Result<(), SemanticError> {
     if let fe::FuncStmt::Emit {
-        value: Node {
-            kind: fe::Expr::Call { func, args },
-            ..
-        },
+        value: Node { kind, .. },
     } = &stmt.kind
     {
-        let event_name = expressions::expr_name_str(func)?;
-
-        if let Some(event) = scope.borrow().contract_event_def(event_name) {
-            context.borrow_mut().add_emit(stmt, event.clone());
-
-            let argument_attributes = args
-                .kind
-                .iter()
-                .map(|arg| expressions::call_arg(Rc::clone(&scope), Rc::clone(&context), arg))
-                .collect::<Result<Vec<_>, _>>()?;
-
-            if fixed_sizes_to_types(event.all_field_types())
-                != expression_attributes_to_types(argument_attributes)
-            {
-                return Err(SemanticError::type_error());
+        return match kind {
+            fe::Expr::Call { func, args } => {
+                let event_name = expressions::expr_name_str(func)?;
+
+                return if let Some(event) = scope.borrow().contract_event_def(event_name) {
+                    context.borrow_mut().add_emit(stmt, event.clone());
+
+                    let argument_attributes = args
+                        .kind
+                        .iter()
+                        .map(|arg| {
+                            expressions::call_arg(Rc::clone(&scope), Rc::clone(&context), arg)
+                        })
+                        .collect::<Result<Vec<_>, _>>()?;
+
+                    if fixed_sizes_to_types(event.all_field_types())
+                        != expression_attributes_to_types(argument_attributes)
+                    {
+                        Err(SemanticError::type_error())
+                    } else {
+                        Ok(())
+                    }
+                } else {
+                    Err(SemanticError::missing_event_definition())
+                };
             }
-        }
-
-        return Ok(());
+            _ => Err(SemanticError::event_invocation_expected()),
```

Comment:
I was wondering if this should rather be handled on the parser level. Emit is defined as 

```
Emit {
        value: Node<Expr<'a>>,
    },
```

So, every expression is possible but I'm not entirely sure if that's what we want. I guess in terms of AST nodes this could also be defined as:

```
Emit {
        value: Node<EventExpr<'a>>,
    },
```

Where `EventExpr` would then already match against the call syntax. So something like `emit foo` (instead of `foo()` could be rejected directly on the parsing stage. The same is true for the initialization of structs I think.

Reply:
I'm pretty indifferent here.

Another idea is that we could go with a slightly more natural syntax like `MyEvent(42, "foo", [0,1,3,4]).emit()`, which would let us remove the `emit` statement entirely.

Reply:
>which would let us remove the `emit` statement entirely

There's something I prefer about the `emit MyEvent()` syntax. It might just be that it stands out more due to the extra keyword being used. That can make it easier to spot in a code base.

Anyway, I'm going to take it as is for now but I'm happy to revisit that in the future.